static/src/javascripts/projects/common/modules/discussion/api.ts

```diff
@@ -0,0 +1,121 @@
+ */
+
+type Id = number | string;
+
+type Comment = {
+	body: string;
+	id?: number;
+	replyTo?: {
+		commentId: string;
+	};
+};
+
+type CommentResponse = {
+	status: 'ok' | 'error';
+	statusCode: number;
+	message: string;
+	errorCode?: string;
+};
+
+type AbuseReport = {
+	categoryId: number;
+	reason?: string;
+	email?: string;
+};
+
+const defaultInitParams: RequestInit = {
+	mode: 'cors',
+	credentials: 'include',
+	headers: {
+		'D2-X-UID': String(config.get('page.discussionD2Uid')),
+		'GU-Client': String(config.get('page.discussionApiClientHeader')),
+	},
+};
+
+export const send = (
+	endpoint: string,
+	method: string,
+	data = '',
+): Promise<CommentResponse> => {
+	if (config.get('switches.enableDiscussionSwitch')) {
+		const url = String(config.get('page.discussionApiUrl')) + endpoint;
+
+		// https://github.com/guardian/discussion-rendering/blob/1e8a7c7fa0b6a4273497111f0dab30f479a107bf/src/lib/api.tsx#L140
+		if (method === 'POST') {
+			const body = new URLSearchParams();
+			body.append('body', data);
+
+			return fetch(url, {
+				...defaultInitParams,
+				method,
+				body: body.toString(),
+				headers: {
+					...defaultInitParams.headers,
+					'Content-Type': 'application/x-www-form-urlencoded',
+				},
+			}).then((resp) => resp.json() as Promise<CommentResponse>);
+		}
+
+		return fetch(url, {
+			...defaultInitParams,
+			method,
+		}).then((resp) => resp.json() as Promise<CommentResponse>);
+	}
+
+	throw new Error('Discussion features have been disabled');
+};
+
+export const postComment = (
+	discussionId: Id,
+	comment: Comment,
+): Promise<CommentResponse> => {
+	const endpoint = `/discussion/${discussionId}/comment${
+		comment.replyTo ? `/${comment.replyTo.commentId}/reply` : ''
+	}`;
+
+	return send(endpoint, 'POST', comment.body);
+};
+
+export const previewComment = (comment: Comment): Promise<CommentResponse> =>
+	send('/comment/preview', 'POST', comment.body);
+
+export const recommendComment = (id: Id): Promise<CommentResponse> =>
+	send(`/comment/${id}/recommend`, 'POST');
+
+export const pickComment = (id: Id): Promise<CommentResponse> =>
+	send(`/comment/${id}/highlight`, 'POST');
+
+export const unPickComment = (id: Id): Promise<CommentResponse> =>
+	send(`/comment/${id}/unhighlight`, 'POST');
+
+export const reportComment = (
+	id: Id,
+	report: AbuseReport,
+): Promise<CommentResponse> => {
+	const data = new URLSearchParams();
+	data.append('categoryId', report.categoryId.toString());
+	report.email && data.append('email', report.email.toString());
+	report.reason && data.append('reason', report.reason);
+
+	const url =
+		String(config.get('page.discussionApiUrl')) +
+		`/comment/${id}/reportAbuse`;
+
+	return fetch(url, {
+		...defaultInitParams,
+		method: 'POST',
+		body: data.toString(),
+		headers: {
+			...defaultInitParams.headers,
+			'Content-Type': 'application/x-www-form-urlencoded',
+		},
+	}).then((resp) => resp.json() as Promise<CommentResponse>);
```

Comment:
why not use the generic `send` method above?

Reply:
Because of the shape of `data`. I’ve tried to handle both cases but couldn’t figure out how to narrow between `Comment | AbuseReport`.

This is also in line with how `discussion-rendering` does it, so I’d rather leave it as-is.